README.md
<h2 align="center">Time to <i>Go Home</i> :clock10:</h2>
<p align="center">
	<a href="https://github.com/fgrosse/go-home/releases"><img src="https://img.shields.io/github/tag/fgrosse/go-home.svg?label=version&color=brightgreen"></a>
	<a href="https://github.com/fgrosse/go-home/blob/master/LICENSE"><img src="https://img.shields.io/badge/license-BSD--3--Clause-blue.svg"></a>
</p>

---

Go Home is a small OpenGL based progress bar widget for your Desktop that
displays for how long you have been working each day. This is helpful for people
who tend to loose track of time and thus do overhours when they actually wanted
to leave home. I made this to learn a bit about simple OpenGL programming using
[the Go programming language][go].

<p align="center">
<img src="assets/screenshot_01.png">
<img src="assets/screenshot_02.png">
</p>

## Installation

### Precompiled binaries

You can find precompiled binaries at the [releases] page of the GitHub
repository.

### From Source

Go Home is packaged using [Go modules][go-modules]. Since this is not a library
but a runnable application within a `main` package you need to clone this
repository first. Typically this should be done outside of the `$GOPATH` or Go
will complain due Modules being enabled.

After you cloned the repo you should make sure to install the external
dependencies (i.e. OpenGL bindings) as explained at the
[GLFW repository][external-deps]. There is a [Makefile](Makefile) to install the
requires libraries on RedHead/Fedora.

Afterwards you simply use `go build` or `go install` and Go will fetch the
correct Go dependencies for you:
 
```bash
$ git clone https://github.com/fgrosse/go-home.git
Cloning into 'go-home'...
remote: Enumerating objects: 122, done.
remote: Counting objects: 100% (122/122), done.
remote: Compressing objects: 100% (69/69), done.
remote: Total 122 (delta 62), reused 109 (delta 49), pack-reused 0
Receiving objects: 100% (122/122), 71.62 KiB | 601.00 KiB/s, done.
Resolving deltas: 100% (62/62), done.

$ cd go-home               
$ make setup
…

$ go install && go-home --debug
2019-06-16 13:16	DEBUG	go-home/config.go:50	Running in debug mode
2019-06-16 13:16	INFO	go-home/config.go:54	Loading configuration	{"path": "/home/fgrosse/.go-home.yml"}
2019-06-16 13:16	INFO	go-home/app.go:56	Starting application	{"config": {"check_in": "2019-06-16 12:18", "work_duration": "8h0m0s", "lunch_duration": "1h0m0s", "day_end": "20:00"}}
```

## Usage

You can start the program without any arguments which will create an
undecorated window that displays when you started the program the first time
today and when its time to go home. The default configuration assumes you are
working 8 hours a day and do 1 hour of lunch break.

As time goes by the progress bar will slowly fill up from green to red. If you 
are working overtime it will start to pulse red to catch your attention. At this
point you should leave home and enjoy your free time with your family and
friends :relaxed:.

### Configuration

Go Home reads configuration from `$HOME/.go-home.yml`. If this file does not
exist on the first start it will be created using sensible default values.
The available options in there should be pretty self explanatory.

## Built With

* [pixel](https://github.com/faiface/pixel) - A hand-crafted 2D game library in Go
* [glfw](https://github.com/go-gl/glfw) - Go bindings for GLFW 3
* [cobra](https://github.com/spf13/cobra) - A Commander for modern Go CLI interactions 
* [zap](https://github.com/uber-go/zap) - Blazing fast, structured, leveled logging in Go
* [pkg/errors](https://github.com/pkg/errors) - Simple error handling primitives
* [gopkg.in/yaml](https://gopkg.in/yaml.v3) - YAML support for the Go language 
* [Glacial Indifference Font](https://fontlibrary.org/en/font/glacial-indifference) - An open source typeface by Alfredo Marco Pradil ([SIL Open Font License](LICENSE_FONT))

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on the code of
conduct and on the process for submitting pull requests to this repository.

## Versioning

This software uses [SemVer] for versioning.
For the versions available, see the [tags on this repository][tags]. 

## Authors

- **Friedrich Große** - *Initial work* - [fgrosse]

See also the list of [contributors] who participated in this project.

## License

This project is licensed under the BSD-3-Clause License - see the [LICENSE](LICENSE) file for details.

[releases]: https://github.com/fgrosse/go-home/releases
[external-deps]: https://github.com/go-gl/glfw/blob/master/README.md
[go]: https://golang.org
[go-modules]: https://github.com/golang/go/wiki/Modules
[SemVer]: http://semver.org
[tags]: https://github.com/fgrosse/go-home/tags
[fgrosse]: https://github.com/fgrosse
[contributors]: https://github.com/github.com/fgrosse/go-home/contributors


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is loosely based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

This project uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- Nothing so far

## [v1.1.0] - 2019-06-16
- Allow moving the window with Shift+ArrowKey

## [v1.0.1] - 2019-06-02
- Fix panic when running in debug mode without Font file

## [v1.0.0] - 2019-06-02
- Initial release.

[Unreleased]: https://github.com/fgrosse/go-home/compare/v1.1.0...HEAD
[v1.1.0]: https://github.com/fgrosse/go-home/releases/tag/v1.1.0
[v1.0.0]: https://github.com/fgrosse/go-home/releases/tag/v1.0.0


app.go
package main

import (
	"image/color"
	"os"
	"time"

	"github.com/faiface/pixel"
	"github.com/faiface/pixel/pixelgl"
	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	"go.uber.org/zap"
)

type App struct {
	*cobra.Command
	logger *zap.Logger
	conf   Config
	win    *pixelgl.Window
	render *Render

	initErr  error
	shutdown bool
	limitFPS bool
}

func NewApp() *App {
	app := &App{Command: &cobra.Command{
		Use: "go-home",
	}}

	app.SilenceUsage = true  // do not output usage in case of an error
	app.SilenceErrors = true // we log them manually in the main function
	app.Command.RunE = app.Run

	var (
		debug  bool
		config string
	)

	flags := app.PersistentFlags()
	flags.StringVar(&config, "config", os.ExpandEnv("$HOME/.go-home.yml"), "config file")
	flags.BoolVar(&debug, "debug", false, "enable debug mode")

	cobra.OnInitialize(app.loadConfig(&debug, &config))

	return app
}

func (app *App) Run(_ *cobra.Command, _ []string) error {
	if app.initErr != nil {
		return app.initErr
	}

	var err error
	app.logger.Info("Starting application", zap.Object("config", app.conf))
	err = app.createWindow(app.conf.UI)
	if err != nil {
		return errors.Wrap(err, "failed to create window")
	}

	app.render, err = NewRender(app.conf)
	if err != nil {
		return errors.Wrap(err, "failed to create renderer")
	}

	app.runLoop()

	err = app.conf.Save()
	if err != nil {
		app.logger.Error("Failed to save configuration on shutdown", zap.Error(err))
	}

	return nil
}

func (app *App) createWindow(conf UIConfig) error {
	width := float64(conf.WindowWidth)
	height := float64(conf.WindowHeight)

	cfg := pixelgl.WindowConfig{
		Title:       "Go Home",
		Bounds:      pixel.R(0, 0, width, height),
		VSync:       true,
		Undecorated: true,
		Resizable:   false,
		AlwaysOnTop: true,
	}

	// TODO: we need GLFW 3.3 where we get the GLFW_TRANSPARENT_FRAMEBUFFER option
	// See: https://www.glfw.org/docs/3.3/window_guide.html#window_hints_wnd

	var err error
	app.win, err = pixelgl.NewWindow(cfg)
	if err != nil {
		return errors.WithStack(err)
	}

	app.win.SetSmooth(true)
	app.win.SetPos(app.conf.UI.WindowPos)
	app.win.Update()

	return nil
}

func (app *App) runLoop() {
	app.limitFPS = true // might be disabled when moving the window via Shift+Arrow
	fps := time.Tick(time.Second / time.Duration(app.conf.UI.FPS))
	last := time.Now()
	for !app.win.Closed() {
		dt := time.Since(last).Seconds()
		last = time.Now()

		// TODO: deal with a new dawn

		app.win.Clear(color.White)
		app.handleInput(app.win, dt)
		app.render.Draw(app.win)
		app.win.Update()

		if app.shutdown {
			return
		}

		if app.limitFPS {
			<-fps
		}
	}
}


input.go
package main

import (
	"math"
	"time"

	"github.com/faiface/pixel/pixelgl"
)

func (app *App) handleInput(win *pixelgl.Window, dt float64) {
	if win.Pressed(pixelgl.KeyLeftShift) {
		app.limitFPS = false

		const speed = 100.0 // pixel per second
		delta := speed * dt
		if win.Pressed(pixelgl.KeyRight) {
			app.conf.UI.WindowPos.X += delta
		} else if win.Pressed(pixelgl.KeyLeft) {
			app.conf.UI.WindowPos.X -= delta
		}

		if win.Pressed(pixelgl.KeyUp) {
			app.conf.UI.WindowPos.Y -= delta
		} else if win.Pressed(pixelgl.KeyDown) {
			app.conf.UI.WindowPos.Y += delta
		}

		currPos := app.win.GetPos()
		if math.Round(currPos.X) != math.Round(app.conf.UI.WindowPos.X) ||
			math.Round(currPos.Y) != math.Round(app.conf.UI.WindowPos.Y) {
			app.win.SetPos(app.conf.UI.WindowPos)
		}

	} else {
		app.limitFPS = true

		if app.conf.Debug {
			const speed = 2 * float64(time.Hour)
			switch {
			case win.Pressed(pixelgl.KeyRight):
				app.render.timeShift += time.Duration(speed * dt)
			case win.Pressed(pixelgl.KeyLeft):
				app.render.timeShift -= time.Duration(speed * dt)
			case win.Pressed(pixelgl.KeyEscape):
				app.shutdown = true
			}
		}
	}
}


logger.go
package main

import (
	"time"

	"github.com/pkg/errors"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

func newLogger(debug bool) *zap.Logger {
	encConf := zap.NewDevelopmentEncoderConfig()
	encConf.EncodeDuration = zapcore.StringDurationEncoder
	encConf.EncodeTime = func(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
		enc.AppendString(t.Format("2006-01-02 15:04"))
	}

	conf := zap.Config{
		Level:            zap.NewAtomicLevelAt(zap.InfoLevel),
		Encoding:         "console",
		EncoderConfig:    encConf,
		DisableCaller:    true,
		OutputPaths:      []string{"stdout"},
		ErrorOutputPaths: []string{"stdout"},
	}

	if debug {
		conf.Level = zap.NewAtomicLevelAt(zap.DebugLevel)
		conf.Development = true
		conf.DisableCaller = false
	}

	logger, err := conf.Build()
	if err != nil {
		panic(errors.Wrap(err, "failed to create logger"))
	}

	return logger
}


render.go
package main

import (
	"image/color"
	"math"
	"time"

	"github.com/faiface/pixel"
	"github.com/faiface/pixel/imdraw"
	"github.com/faiface/pixel/text"
	"github.com/fgrosse/go-home/assets"
	"github.com/golang/freetype/truetype"
	"github.com/pkg/errors"
	"golang.org/x/image/colornames"
	"golang.org/x/image/font"
)

type Render struct {
	Width, Height     float64
	CheckIn           time.Time
	CheckOut          time.Time
	EOD               time.Time
	Atlas             *text.Atlas
	MarkerColor       color.Color
	BorderColor       color.Color
	ShowRemainingTime bool          // show how much time is left instead of the current time
	timeShift         time.Duration // for debugging
}

func NewRender(conf Config) (*Render, error) {
	fnt, err := loadTTF("/assets/GlacialIndifference-Regular.ttf", 12)
	if err != nil {
		return nil, errors.Wrap(err, "failed to load font")
	}

	return &Render{
		Width:             float64(conf.UI.WindowWidth),
		Height:            float64(conf.UI.WindowHeight),
		CheckIn:           conf.CheckIn,
		CheckOut:          conf.CheckOut,
		EOD:               conf.EndOfDay,
		Atlas:             text.NewAtlas(fnt, text.ASCII),
		MarkerColor:       colornames.Mediumblue,
		ShowRemainingTime: conf.UI.ShowRemainingTime,
	}, nil
}

func loadTTF(path string, size float64) (font.Face, error) {
	bytes := assets.FSMustByte(false, path)
	fnt, err := truetype.Parse(bytes)
	if err != nil {
		return nil, err
	}

	return truetype.NewFace(fnt, &truetype.Options{
		Size:              size,
		GlyphCacheEntries: 1,
	}), nil
}

func (r *Render) Draw(t pixel.Target) {
	now := time.Now().Add(r.timeShift)
	progress := r.progress(now)

	markerTxt := r.markerText(progress, now)
	checkoutTxt := r.checkoutText(now)

	r.drawGradient(t, progress)
	r.drawCells(t, markerTxt, checkoutTxt)
	r.drawTargetMarker(t, progress, markerTxt)
	r.drawCurrentMarker(t, progress, now, markerTxt)
	r.drawText(t, now, markerTxt, checkoutTxt)
	r.drawRectangle(t, now)
}

func (r *Render) drawGradient(t pixel.Target, progress float64) {
	leftColor := pixel.RGB(0, 1, 0)
	rightColor := pixel.RGB(progress, 1-progress, 0)

	rect := imdraw.New(nil)
	rect.Color = leftColor
	rect.Push(pixel.V(0, 0))
	rect.Push(pixel.V(0, r.Height))

	rect.Color = rightColor
	rect.Push(pixel.V(r.Width*progress, r.Height))
	rect.Push(pixel.V(r.Width*progress, 0))
	rect.Polygon(0)
	rect.Draw(t)

	// draw gray scale gradient
	rect = imdraw.New(nil)
	gray := pixel.RGB(0.333, 0.333, 0.333)
	rect.Color = gray.Scaled(progress)
	rect.Push(pixel.V(r.Width*progress+1, 0))
	rect.Push(pixel.V(r.Width*progress+1, r.Height))

	rect.Color = gray
	rect.Push(pixel.V(r.Width, r.Height))
	rect.Push(pixel.V(r.Width, 0))
	rect.Polygon(0)
	rect.Draw(t)
}

func (r *Render) drawCells(t pixel.Target, markerTxt, checkoutTxt *text.Text) {
	numCells := 9
	txtBounds := markerTxt.Bounds()
	checkoutBounds := checkoutTxt.Bounds()
	for i := 0; i < numCells; i++ {
		x := r.Width * float64(i) / float64(numCells)
		line := imdraw.New(nil)
		line.Color = color.Black

		v := pixel.V(x, r.Height/2)
		if txtBounds.Contains(v) || checkoutBounds.Contains(v) {
			col := pixel.ToRGBA(line.Color)
			col.A = 0.1
			line.Color = col
		}

		line.Push(pixel.V(x, 0))
		line.Push(pixel.V(x, r.Height))
		line.Line(1)
		line.Draw(t)
	}
}

func (r *Render) drawTargetMarker(t pixel.Target, progress float64, markerTxt *text.Text) {
	posX := r.position(r.CheckOut)

	imd := imdraw.New(nil)
	col := pixel.RGB(0, 0, 0)
	if markerTxt.Bounds().Contains(pixel.V(posX, r.Height/2)) {
		col.A = 0.1
	}

	imd.Color = col
	imd.Push(pixel.V(posX, 2))
	imd.Push(pixel.V(posX, r.Height-2))
	imd.Line(3)
	imd.Draw(t)
}

func (r *Render) drawCurrentMarker(t pixel.Target, progress float64, now time.Time, txt *text.Text) {
	posX := progress * r.Width
	col := r.MarkerColor
	if now.After(r.CheckOut) {
		col = r.BorderColor
	}

	imd := imdraw.New(nil)
	imd.Color = col
	imd.Push(pixel.V(posX, 2))
	imd.Push(pixel.V(posX, r.Height-2))
	imd.Line(2)
	imd.Draw(t)

	imd = imdraw.New(nil)
	imd.Color = col
	imd.Push(pixel.V(posX-5, 2))
	imd.Push(pixel.V(posX+1+5, 2))
	imd.Push(pixel.V(posX+1, 5+2))
	imd.Push(pixel.V(posX, 5+2))
	imd.Polygon(0)
	imd.Draw(t)

	imd = imdraw.New(nil)
	imd.Color = col
	imd.Push(pixel.V(posX-5, r.Height-2))
	imd.Push(pixel.V(posX+1+5, r.Height-2))
	imd.Push(pixel.V(posX+1, r.Height-5-2))
	imd.Push(pixel.V(posX, r.Height-5-2))
	imd.Polygon(0)
	imd.Draw(t)
}

func (r *Render) markerText(progress float64, now time.Time) *text.Text {
	posX := progress * r.Width
	txt := text.New(pixel.V(posX+5, 4), r.Atlas)
	if progress > 0.7 {
		txt.Color = color.White
	} else {
		txt.Color = color.Black
	}

	if r.ShowRemainingTime {
		remaining := r.CheckOut.Sub(now).Round(time.Minute)
		s := remaining.String()
		s = s[:len(s)-2] // strip away trailing seconds
		txt.WriteString(s)
	} else {
		txt.WriteString(now.Format("15:04"))
	}

	return txt
}

func (r *Render) checkoutText(now time.Time) *text.Text {
	txt := text.New(pixel.V(4+r.position(r.CheckOut), 4), r.Atlas)
	if now.Before(r.CheckOut) {
		txt.Color = color.White
	} else {
		txt.Color = color.Black
	}

	txt.WriteString(r.CheckOut.Format("15:04"))
	return txt
}

func (r *Render) drawRectangle(t pixel.Target, now time.Time) {
	var borderWidth float64 = 2

	rect := imdraw.New(nil)
	if now.Before(r.CheckOut) {
		r.BorderColor = color.Black
	} else {
		// Pulsating red border
		totalMillis := int(float64(now.UnixNano()) / float64(time.Millisecond))
		scale := (1 + math.Sin(float64(totalMillis)/500)) / 2
		r.BorderColor = pixel.ToRGBA(colornames.Red).Mul(pixel.RGB(scale, scale, scale))
		borderWidth = 4
	}

	rect.Color = r.BorderColor
	rect.EndShape = imdraw.RoundEndShape
	rect.Push(pixel.V(1, 1))
	rect.Push(pixel.V(1, r.Height-1))
	rect.Push(pixel.V(r.Width-1, r.Height-1))
	rect.Push(pixel.V(r.Width-1, 1))
	rect.Push(pixel.V(1, 1))
	rect.Line(borderWidth)
	rect.Draw(t)
}

func (r *Render) drawText(t pixel.Target, now time.Time, markerTxt, checkoutTxt *text.Text) {
	txt := text.New(pixel.V(4, 4), r.Atlas)
	txt.Color = color.Black
	txt.WriteString(r.CheckIn.Format("15:04"))
	bounds := txt.Bounds()

	shift := r.Height/2 - bounds.Size().Y/2 - 1
	m := pixel.IM.Moved(pixel.V(0, shift))

	if bounds.Intersect(markerTxt.Bounds()) == pixel.ZR {
		txt.Draw(t, m)
	}

	if checkoutTxt.Bounds().Intersect(markerTxt.Bounds()) == pixel.ZR {
		checkoutTxt.Draw(t, m)
	}

	markerTxt.Draw(t, m)
}

func (r *Render) position(t time.Time) float64 {
	return r.Width * r.progress(t)
}

func (r *Render) progress(now time.Time) float64 {
	left := float64(r.CheckIn.Unix())
	right := float64(r.EOD.Unix())
	nowUnix := float64(now.Unix())

	totalSec := right - left
	diffSec := nowUnix - left
	return diffSec / totalSec
}


config.go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"math"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/faiface/pixel"
	"github.com/pkg/errors"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
	"gopkg.in/yaml.v3"
)

type Config struct {
	CheckIn  time.Time `yaml:"check_in"`
	CheckOut time.Time `yaml:"-"`
	EndOfDay time.Time `yaml:"-"`

	WorkDuration  time.Duration `yaml:"work_duration"`
	LunchDuration time.Duration `yaml:"lunch_duration"`
	DayEnd        ClockTime     `yaml:"day_end"`

	UI    UIConfig `yaml:"ui"`
	Debug bool     `yaml:"-"`

	path string `yaml:"-"`
}

type UIConfig struct {
	FPS               int       `yaml:"fps"`
	WindowWidth       int       `yaml:"width"`
	WindowHeight      int       `yaml:"height"`
	WindowPos         pixel.Vec `yaml:"pos"`
	ShowRemainingTime bool      `yaml:"show_remaining_time"`
}

func (app *App) loadConfig(debug *bool, path *string) func() {
	return func() {
		app.logger = newLogger(*debug)
		if app.initErr != nil {
			return
		}

		app.logger.Debug("Running in debug mode")

		var r io.Reader
		if f, err := os.Open(*path); err == nil {
			app.logger.Info("Loading configuration", zap.String("path", *path))
			r = f
			defer f.Close()
		} else if os.IsNotExist(err) {
			app.logger.Info("Configuration file not found. Creating new file", zap.String("path", *path))
		} else {
			app.initErr = errors.Wrap(err, "failed to open config file")
			return
		}

		app.conf, app.initErr = LoadConfig(r, app.logger, *path, *debug)
		if app.initErr != nil {
			return
		}

		app.initErr = app.conf.Save()
	}
}

func LoadConfig(r io.Reader, logger *zap.Logger, path string, debug bool) (Config, error) {
	conf := Config{path: path}
	if r != nil {
		dec := yaml.NewDecoder(r)
		dec.KnownFields(true)
		err := dec.Decode(&conf)
		if err != nil {
			return conf, errors.Wrap(err, "failed to decode config")
		}
	}

	if conf.UI.WindowWidth == 0 {
		conf.UI.WindowWidth = 512
	}
	if conf.UI.WindowHeight == 0 {
		conf.UI.WindowHeight = 32
	}

	if conf.UI.WindowPos.X == 0 && conf.UI.WindowPos.Y == 0 {
		var displayWidth float64 = 1920 // TODO: make dynamic
		conf.UI.WindowPos = pixel.Vec{
			X: displayWidth/2 - float64(conf.UI.WindowWidth)/2,
			Y: 35,
		}
	}

	if conf.WorkDuration == 0 {
		conf.WorkDuration = 8 * time.Hour
	}
	if conf.LunchDuration == 0 {
		conf.LunchDuration = 1 * time.Hour
	}
	if conf.DayEnd.Hour == 0 {
		conf.DayEnd = ClockTime{Hour: 20, Minute: 00}
	}
	if conf.UI.FPS == 0 {
		conf.UI.FPS = 10
	}
	if conf.CheckIn.IsZero() || isDifferentDay(conf.CheckIn, time.Now()) {
		conf.CheckIn = time.Now()
		logger.Info("Detected start of new day", zap.String("date", conf.CheckIn.Format("2006-01-02")))
	}

	conf.CheckIn = conf.CheckIn.Round(time.Second)
	conf.CheckOut = conf.CheckIn.Add(conf.WorkDuration).Add(conf.LunchDuration)
	conf.EndOfDay = conf.DayEnd.Time(conf.CheckIn)
	conf.Debug = debug

	return conf, nil
}

func isDifferentDay(a, b time.Time) bool {
	yearA, monthA, dayA := a.Date()
	yearB, monthB, dayB := b.Date()
	return yearA != yearB || monthA != monthB || dayA != dayB
}

func (conf Config) Save() error {
	conf.UI.WindowPos.X = math.Round(conf.UI.WindowPos.X)
	conf.UI.WindowPos.Y = math.Round(conf.UI.WindowPos.Y)

	data, err := yaml.Marshal(conf)
	if err != nil {
		return errors.Wrap(err, "failed to encode config as YAML")
	}

	err = ioutil.WriteFile(conf.path, data, 0666)
	if err != nil {
		return errors.Wrap(err, "failed to save config")
	}

	return nil
}

func (conf Config) MarshalLogObject(enc zapcore.ObjectEncoder) error {
	if conf.CheckIn.IsZero() {
		enc.AddString("check_in", "-")
	} else {
		enc.AddString("check_in", conf.CheckIn.Format("2006-01-02 15:04"))
	}

	enc.AddDuration("work_duration", conf.WorkDuration)
	enc.AddDuration("lunch_duration", conf.LunchDuration)
	enc.AddString("day_end", conf.DayEnd.String())

	return nil
}

type ClockTime struct {
	Hour, Minute int
}

func (t *ClockTime) UnmarshalText(text []byte) error {
	parts := strings.Split(string(text), ":")
	if len(parts) != 2 {
		return errors.New(`ClockTime string is not formatted as "hh:mm"`)
	}

	var err error
	t.Hour, err = strconv.Atoi(parts[0])
	if err != nil {
		return errors.Errorf("hour part is not an integer")
	}

	t.Minute, err = strconv.Atoi(parts[1])
	if err != nil {
		return errors.Errorf("minute part is not an integer")
	}

	if t.Hour < 0 || t.Hour > 24 {
		return errors.Errorf("invalid hour")
	}

	if t.Minute < 0 || t.Minute > 59 {
		return errors.Errorf("invalid minute")
	}

	return nil
}

func (t ClockTime) Time(ref time.Time) time.Time {
	year, month, day := ref.Date()
	return time.Date(year, month, day, t.Hour, t.Minute, 0, 0, ref.Location())
}

func (t ClockTime) String() string {
	return fmt.Sprintf("%02d:%02d", t.Hour, t.Minute)
}

func (t ClockTime) MarshalYAML() (interface{}, error) {
	return t.String(), nil
}
